sefaria/helper/schema.py

```diff
@@ -617,10 +618,37 @@ def construct_query(attribute, queries):
         generic_rewrite(HistorySet(construct_query('old.refs', identifier)), attr_name='old', sub_attr_name='refs')
 
 
-def migrate_to_complex_structure(title, schema, mappings, rewriter, needs_rewrite):
-    #print title
-    #print mappings
-    #print json.dumps(schema)
+def migrate_to_complex_structure(title, schema, mappings):
+    """
+    Converts book that is simple structure to complex.
+    :param title: title of book
+    :param schema: the new complex structure schema, must be JSON
+    :param mappings: a dictionary mapping references from simple structure to references in complex structure
+                    For example:
+        mappings = {"Midrash Tanchuma 1:1": "Midrash Tanchuma, Bereshit",
+                    "Midrash Tanchuma 1:2": "Midrash Tanchuma, Noach",
+                    ...
+                    "Midrash Tanchuma 2:1": "Midrash Tanchuma, Shemot"}
+    :return:
+    """
+    def needs_rewrite(ref, *args):
+        try:
+            return Ref(ref).index.title == title
+        except InputError:
+            return False
+
+    def rewriter(ref):
+        ref = Ref(ref)
+        for old_ref in mappings:
+            old_ref = Ref(old_ref)
+            if old_ref.contains(ref):
```

Comment:
I don't think this will correctly handle ranged references.  
If the original refs is of the form "Text Name 1:2-4" for example.

Some of these will be in one jagged array in the target schema, and thus convertible.
Some will bridge over two schema nodes in the target, and will need to be truncated (or deleted.)